File: C,C++/Project4/LinkedList.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   LinkedList.cpp
 * Author: eriko
 * 
 * Created on November 1, 2017, 12:43 PM
 */
#include <cstdlib>
#include <string>
#include "Value.h"
#include "Comparator.h"
#include "LinkedList.h"


using namespace std;

bool BasicComparator::compare(Value *a, Value *b){
    return a->getValue()>b->getValue();
}


LinkedList::LinkedList(Value* v, Comparator* c){
    value = v;
    comparator = c;
}

LinkedList::LinkedList(Comparator * c){
    comparator = c;
}

LinkedList::LinkedList(){
    comparator = new BasicComparator();
}

bool LinkedList::hasNext(){
    return next!=nullptr;
}

bool LinkedList::hasPrev(){
    return prev!=nullptr;
}

void LinkedList::setNext(LinkedList* node){ //sets the next node of the linked list and sets the previous node for the new node to the current node
    next = node;
    if(node!=nullptr)
        node->prev = this;
}

void LinkedList::setPrev(LinkedList *node){ //sets the previous node of the current node to the given node sets the next node of the new node to the current node
    prev = node;
    if(node!=nullptr)
        node->next = this;
}

void LinkedList::add(Value * value){    //adds a new value to the linked list in sorted order
    if(this->value==nullptr)                                //if empty set value to given value
        this->value = value;
    else{
        if(comparator->compare(value,this->value)){         
            if(hasNext())                                   //if current node has a next node, go to the next node
                next->add(value);
            else
                setNext(new LinkedList(value,comparator));  //else set next node to new node
        }
        else{
            push(value);                                    //else current node>value insert value into current position
        }
    }
}

void LinkedList::push(Value * value){   //inserts value into current position
    if(value!=nullptr){                 //if no value exists
        if(hasNext())                   //and current node has a next node
            next->push(this->value);    //push current node to next node
        else
            setNext(new LinkedList(this->value,comparator));    //else set the next node to a copy of the current node
    }
    this->value = value;    //set the value of the current node to the given value
}

void LinkedList::append(Value * value){
    if(hasNext())
        this->next->append(value);
    else{
        if(this->value==nullptr)
            this->value = value;
        else
            setNext(new LinkedList(value,comparator));
    }
}

Value* LinkedList::pop(){   //return and remove the first value of the list
    Value* value = this->value; //get current value
    deleteSelf();               //delete current node
    return value;               //return value
}

LinkedList* LinkedList::get(string value){  //returns the node containing the given value
    if(this->value->equal(value))       //compares value against current node
        return this;
    if(hasNext())                       //if current node has next node
        return this->next->get(value);  //check next node
    return nullptr;                     //if it does not return null pointer
}

LinkedList* LinkedList::get(int index){     //returns the node at given index
    if(index==0)                //if index = 0 return current node
        return this;
    if(hasNext())               //if has next node
        return next->get(index-1);  //check next node while decrementing index
    return nullptr;             //else return null pointer
}

void LinkedList::remove(string value){  //remove node containing given value
    if(this->value->equal(value))   //if current node contains given value 
        deleteSelf();               //delete current node
    else if(hasNext())              //if has next node 
        next->remove(value);        //check next node        
}

void LinkedList::remove(int index){     //remove node at given index
    if(index==0)            //if index = 0 
        deleteSelf();       //delete current node
    else if(hasNext())      //if has next node 
        next->remove(index-1);  //check next node while decrementing index
}

string LinkedList::toString(){  //converts list to string
    string res = "";
    if(value!=nullptr)                      //if current value is not null
        res+=value->getValue();             //add value to string
    if(hasNext())                           //if current has next node
        return res+=","+next->toString();   //add ',' to string and get string of next node
    return res;                             //return result
}

int LinkedList::size(){ //gets the size of the list
    if(hasNext())               //if current node has next node
        return next->size()+1;  //get size of next node and add 1
    if(value!=nullptr)          //if value is not null
        return 1;               
    return 0;
}

bool LinkedList::isEmpty(){
    return size()==0;
}

void LinkedList::deleteSelf(){  //deletes current node from list
    if(hasPrev())                           //if current node has previous node
        prev->setNext(next);                //chain previous node to next node
    else{
        if(hasNext()){                      //if has next
            this->value = next->value;      //replace current value with next value
            this->next = this->next->next;  //replace current's next node with the next node of current's next node
        }
        else
            this->value = nullptr;          //otherwise set value to null pointer
    }
}

File: Java/School/CS 2336/Homework8/src/homework8/LinkedList.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package homework8;

import java.util.Comparator;

/**
 *
 * @author eriko
 */
public class LinkedList<V>{
    LinkedList<V> previous;
    LinkedList<V> next;
    V value;    
    Comparator comparator;
    public LinkedList(Comparator comparator){
        this.comparator = comparator;
    }
    
    public LinkedList(V value, Comparator comparator){
        this.value = value;
        this.comparator = comparator;
    }
    
    public void setPrevious(LinkedList parent){
        previous = parent;
        if(parent!=null)
            parent.next = this;
    }
    
    public LinkedList getPrevious(){
        return previous;
    }
    
    public void setNext(LinkedList child){
        next = child;
        if(child!=null)
            child.previous = this;
    }
    
    public LinkedList getNext(){
        return next;
    }
    
    public V getValue(){
        return value;
    }
    
    public void setValue(V value){
        this.value = value;
    }
    
    public boolean hasNext(){
        return next!=null;
    }
    
    public boolean hasPrevious(){
        return previous!=null;
    }
    
    public boolean hasValue(){
        return value!=null;
    }
    
    public void add(LinkedList<V> list){
            if(hasValue()){
                if(comparator.compare(value, list.value)>0){
                    V temp = value;
                    value = list.getValue();
                    list.setValue(temp);
                    list.setNext(next);
                    setNext(list);
                }
                else{
                    if(hasNext())
                        next.add(list);
                    else
                        setNext(list);
                }    
            }
            else
                value = list.getValue();
        
            
    }
    
    public void add(V value){
        add(new LinkedList(value,comparator));
    }
    
    public void insert(LinkedList<V> list, int index){
        if(index==0){
            add(list);
        }
        else{
            if(hasNext())
                next.insert(list, index-1);
            else
                setNext(list);
        }
    }
    
    public void insert(V value, int index){
        insert(new LinkedList(value,comparator),index);
    }
    
    public boolean delete(V value){
        if(this.value.equals(value)){
            if(hasPrevious())
                previous.setNext(next);
            else{
                if(hasNext()){
                    this.value = next.getValue();
                    this.next = next.getNext();
                }
                else{
                    this.value = null;
                    this.next = null;
                }
            }
            return true;
        }
        else{
            if(hasNext())
                return next.delete(value);
            else
                return false;
        }            
    }
    
    public boolean delete(int index){
        if(index==0){
            if(hasPrevious())
                previous.setNext(next);
            else{
                if(hasNext()){
                    this.value = next.getValue();
                    this.next = next.getNext();
                }
                else{
                    this.value = null;
                    this.next = null;
                }
            }
            return true;
        }
        else{
            if(hasNext())
                return next.delete(index-1);
            else
                return false;
        }      
    }
    
    public V get(int index){
        if(index==0)
            return value;
        if(hasNext())
            return next.get(index-1);
        return null;
    }
    
    public boolean contains(V value){
        if(hasValue() && this.value.equals(value))
            return true;
        if(hasNext())
            return next.contains(value);
        return false;
    }
    
    public String toString(){
        String result = hasPrevious() ? "" : "["; 
        result += hasValue() ? value.toString() : "null";
        return result += hasNext() ? ", "+next.toString() : "]";
    }
    
    public int size(){
        return 1 + (hasNext() ? next.size() : 0);
    }
    
}
